### PR TITLE
Introduce a type-safe API for single-attribute subscriptions

### LINKED ISSUE #11641: Expose Subscription liveness callbacks to the application layer
#### Problem
There doesn't seem to be a way for an application to sign up for `SubscriptionTerminated` notifications from the SDK. 

This means that if a subscription expires, no notification is sent to the application to attempt a retry, for example. 

 #### Proposed Solution
Expose the IMDelegate APIs to the application layer.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
#### Problem
Our subscription APIs are not very type-safe and can't represent all Matter types.

#### Change overview
Introduce an API that subscribes to a single attribute that can do both.

#### Testing
Modified chip-tool command line to use the new API and tested using that.

### CODE REVIEW

**Review by yunhanw-google — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

File: src/controller/ReadInteraction.h [hunk 1]
```diff
@@ -25,30 +25,24 @@
...
- * To avoid instantiating all the complicated code on a per-attribute basis, we
- * have a helper that's just templated on the type.
- */
 template <typename DecodableAttributeType>
-CHIP_ERROR ReadAttribute(Messaging::ExchangeManager * aExchangeMgr, const SessionHandle sessionHandle, EndpointId endpointId,
-                         ClusterId clusterId, AttributeId attributeId,
-                         typename TypedReadAttributeCallback<DecodableAttributeType>::OnSuccessCallbackType onSuccessCb,
-                         typename TypedReadAttributeCallback<DecodableAttributeType>::OnErrorCallbackType onErrorCb)
+CHIP_ERROR ReportAttribute(Messaging::ExchangeManager * exchangeMgr, EndpointId endpointId, ClusterId clusterId,
+                           AttributeId attributeId, ReportParams<DecodableAttributeType> && readParams)
```
Comment (yunhanw-google) [inline src/controller/ReadInteraction.h:43]: Should also update the ReadEvent as well?

Comment (bzbarsky-apple) [reply src/controller/ReadInteraction.h:43]: I'd like to keep these PRs somewhat scoped: this is needed to replace existing broken code; the ReadEvent case would be new code that importantly would not be exercised by anything now and so hard to test.  I'm happy to do it, but in a followup once we have use cases for it.  Assuming they appear.

**Comment by bzbarsky-apple** [comment]
@msandstedt @saurabhst @jmartinez-silabs @jepenven-silabs @LuDuda @Damian-Nordic @chrisdecenzo Please take a look?

File: src/app/ReadClient.h [hunk 2]
```diff
@@ -110,15 +112,15 @@ class ReadClient : public Messaging::ExchangeDelegate
         /**
          * OnSubscriptionEstablished will be called when a subscription is established for the given subscription transaction.
          *
-         * The ReadClient object MUST continue to exist after this call is completed. The application shall wait until it
-         * receives an OnDone call before it shuts down the object.
+         * The ReadClient object MUST continue to exist after this call is completed. The application shall not shut down the
+         * object until after this call has returned, and shall not delete the object until OnDone is called.
```
[the remark this answers is not in the document]
Comment (mrjerryjohns) [reply src/app/ReadClient.h:116]: More-over, ReadClient has still yet to be converted to being application-allocated, so deletion as a concept does not yet apply here.

File: src/controller/CHIPCluster.h [hunk 3]
```diff
@@ -205,6 +206,56 @@ class DLL_EXPORT ClusterBase
...
+        return SubscribeAttribute<typename AttributeInfo::DecodableType, typename AttributeInfo::DecodableArgType>(
+            context, AttributeInfo::GetClusterId(), AttributeInfo::GetAttributeId(), reportCb, failureCb, minIntervalFloorSeconds,
+            maxIntervalCeilingSeconds, subscriptionEstablishedCb);
+    }
+
+    template <typename DecodableType, typename DecodableArgType>
+    CHIP_ERROR SubscribeAttribute(void * context, ClusterId clusterId, AttributeId attributeId,
+                                  ReadResponseSuccessCallback<DecodableArgType> reportCb, ReadResponseFailureCallback failureCb,
+                                  uint16_t minIntervalFloorSeconds, uint16_t maxIntervalCeilingSeconds,
+                                  SubscriptionEstablishedCallback subscriptionEstablishedCb = nullptr)
```
Comment (mrjerryjohns) [inline src/controller/CHIPCluster.h:228]: How do you terminate a subscription after it's been established with such an API? If you were to repeatedly call this with new subscription requests, would that not leak ReadClient's to the point that subscriptions could not be further established?

**Comment by andy31415** [comment]
fast track: change created and approved by domain owner, up sufficient time for all timezones to have a chance to review.

**Review by mrjerryjohns — CHANGES_REQUESTED** [review]

File: src/app/ReadClient.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/ReadClient.h:116]: That's fair; I think I had lost track of these things being pool allocated (which we want to change) when I was changing the comments here...   

I think calling Shutdown "not from inside one of the handlers" is always fine.  Doing it from inside the handlers can get us into various corrupted states, though.....

File: src/controller/CHIPCluster.h (hunk 3 above)
Comment (bzbarsky-apple) [reply src/controller/CHIPCluster.h:228]: > How do you terminate a subscription after it's been established with such an API?

You don't, just like the API it's a replacement for.

The best you can do is create a new subscription to the same server, which kills the first one on the server and then eventually that gets notified and the ReadClient shut down.

That said, we might in fact want to change this API to hand out the ReadClient so you can shut down the subscription.  We want to get our ReadClient lifetime/handle story straight first, right?

**Pull request merged**